### PR TITLE
feat(mme): NGAP Unit Test CMake changes and addition of new cases

### PR DESCRIPTION
**Pull request by panyogesh**
Changes:
  1.  Cmake changes for ngap unit testing directory.
  2.  Renamed the existing files to show the coverage area.
  3.  Added new cases for handling new connection
 
Test:
         1. Unit Testing (test_oai)
             Test Logs :
            UnitTest  Logs
               6/18 Test  #6: test_ngap .........................   Passed    0.19 sec
   
      Call setup Logs
        MME Logs:
         000633 Wed Oct 06 02:20:41 2021 7F6B155E9700 DEBUG NGAP   tasks/ngap/ngap_client_servicer.:0044       
          Is Mock Enabled 0

      UERANSIM Logs:
          [2021-10-06 02:06:13.973] [ngap] [info] PDU session resource(s) setup for UE[1] count[1]
          [2021-10-06 02:06:13.997] [app] [info] Connection setup for PDU session[1] is successful, TUN 
          interface[uesimtun0, 192.168.128.254] is up.

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by ulaskozat — CHANGES_REQUESTED** [review]
This adds a third pattern in unit testing. Currently we use gmock and `#if !MME_UNIT_TEST` patterns. I think we should not have a runtime logic to check whether this is unit testing or not. cc @ardzoht @markjen

**Comment by panyogesh** [comment]
> This adds a third pattern in unit testing. Currently we use gmock and `#if !MME_UNIT_TEST` patterns. I think we should not have a runtime logic to check whether this is unit testing or not. cc @ardzoht @markjen

Sure.. I will change it to compile time.

File: lte/gateway/c/core/oai/test/ngap/test_ngap_flows.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      0x00, 0x01, 0x00, 0x26, 0x00, 0x1a, 0x19, 0x7e, 0x00, 0x41, 0x79,
+      0x00, 0x0d, 0x01, 0x22, 0x62, 0x54, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x01, 0x2e, 0x04, 0xf0, 0xf0, 0xf0, 0xf0, 0x00,
+      0x79, 0x00, 0x13, 0x48, 0x22, 0x42, 0x65, 0x00, 0x00, 0x00, 0x01,
+      0x00, 0x22, 0x42, 0x65, 0x00, 0x00, 0x01, 0xe4, 0xf7, 0x04, 0x44,
+      0x00, 0x5a, 0x40, 0x01, 0x18, 0x00, 0x70, 0x40, 0x01, 0x00};
+
+  itti_init(
+      TASK_MAX, THREAD_MAX, MESSAGES_ID_MAX, tasks_info, messages_info, NULL,
+      NULL);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/ngap/test_ngap_flows.cpp:42]: is this still needed if there are no actual itti messages being passed among the threads?

File: lte/gateway/c/core/oai/tasks/ngap/ngap_amf.c [hunk 2]
```diff
@@ -377,5 +379,5 @@ status_code_e ngap_send_msg_to_task(
       itti_get_task_name(destination_task_id), ITTI_MSG_ID(message),
       ITTI_MSG_NAME(message));
 
-  return send_msg_to_task(task_zmq_ctx_p, destination_task_id, message);
+  return send_message_to_amf(task_zmq_ctx_p, destination_task_id, message);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/ngap/ngap_amf.c]: why not instead move `ngap_send_msg_to_task` to c++ file and have the content of `send_message_to_amf` in there?

File: lte/gateway/c/core/oai/tasks/ngap/ngap_client_enforcer.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#include <string>
+
+using magma5g::NGAPClientServicer;
+
+status_code_e send_message_to_amf(
+    task_zmq_ctx_t* task_zmq_ctx_p, task_id_t destination_task_id,
+    MessageDef* message) {
+  return (NGAPClientServicer::getInstance().send_message_to_amf(
+      task_zmq_ctx_p, destination_task_id, message));
+}
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/ngap/ngap_client_enforcer.cpp]: why can't this be merged with `ngap_client_servicer.cpp`?

File: lte/gateway/c/core/oai/tasks/ngap/include/ngap_client_enforcer.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+#include "intertask_interface.h"
+#include "itti_free_defined_msg.h"
+
+status_code_e send_message_to_amf(
+    task_zmq_ctx_t* task_zmq_ctx_p, task_id_t destination_task_id,
+    MessageDef* message);
+
+#ifdef __cplusplus
+}
+#endif
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/ngap/include/ngap_client_enforcer.h]: why can't this be merged with `ngap_client_servicer.h`?

File: lte/gateway/c/core/oai/test/ngap/test_ngap_flows.cpp (hunk 1 above)
Comment (panyogesh) [reply lte/gateway/c/core/oai/test/ngap/test_ngap_flows.cpp:42]: itti_init() is needed as it fills the MessageDef struct. For instance in func _ngap_amf_itti_ngap_initial_ue_message()_ a new message is allocated (NGAP_INITIAL_UE_MESSAGE). Without itti_init indexing can't be done for _NGAP_INITIAL_UE_MESSAGE_ and hence will result in crash.

File: lte/gateway/c/core/oai/tasks/ngap/ngap_amf.c (hunk 2 above)
Comment (panyogesh) [reply lte/gateway/c/core/oai/tasks/ngap/ngap_amf.c]: Comments addressed.  Removed the client_enforcer layer along with send_message_to_amf() funciton.

File: lte/gateway/c/core/oai/tasks/ngap/ngap_client_enforcer.cpp (hunk 3 above)
Comment (panyogesh) [reply lte/gateway/c/core/oai/tasks/ngap/ngap_client_enforcer.cpp]: Comments addressed.  Removed the client_enforcer layer.

File: lte/gateway/c/core/oai/tasks/ngap/include/ngap_client_enforcer.h (hunk 4 above)
Comment (panyogesh) [reply lte/gateway/c/core/oai/tasks/ngap/include/ngap_client_enforcer.h]: Comments addressed.  Removed the client_enforcer layer.

File: lte/gateway/c/core/oai/test/ngap/test_ngap_flows.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      .outstreams    = 2,
+      .assoc_id      = 3,
+      .ran_cp_ipaddr = ran_cp_ipaddr,
+  };
+
+  EXPECT_EQ(ngap_handle_new_association(state, &p), RETURNok);
+  EXPECT_EQ(state->gnbs.num_elements, 1);
+
+  Ngap_NGAP_PDU_t decoded_pdu = {};
+  unsigned short length =
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/ngap/test_ngap_flows.cpp]: legit github action comment

**Review by ulaskozat — APPROVED** [review]
LGTM

**Comment by ulaskozat** [comment]
@panyogesh there is one minor nit. please address it and then mark this as ready to merge.

File: lte/gateway/c/core/oai/test/ngap/test_ngap_flows.cpp (hunk 5 above)
Comment (panyogesh) [reply lte/gateway/c/core/oai/test/ngap/test_ngap_flows.cpp]: Comments addressed.

**Review by ulaskozat — APPROVED** [review]
LGTM

**Pull request merged**